### PR TITLE
Add Local Time-Step Capability to MPAS-Ocean

### PR DESCRIPTION
**Pull request by gcapodag**
Included a local time-stepping (LTS) method for the barotropic ocean that uses an operator splitting approach of the fast and slow tendencies. The slow tendencies are advanced with an explicit Euler step whereas the fast ones are advanced with an LTS method of order three based on a strong stability preserving RK method [3]. The scheme is overall first order accurate in time. The tendencies for the fast modes are implemented ad-hoc for LTS in a way that allows us to avoid the use of multiple blocks as in our initial implementation [1,2].

At this point, LTS is being added as an MPAS-Standalone feature. LTS was created for ICOM-related coastal inundation studies in ocean standalone mode, and only runs efficiently for single layer simulations. It doesn’t make sense to use LTS for layered-ocean simulations, which includes all E3SM climate simulations. LTS only provides substantial speed-ups for single layer simulations because it does not yet take advantage of barotropic/baroclinic splitting. For layered-ocean simulations, the split time-stepping schemes are more efficient.

References:
[1] https://www.sciencedirect.com/science/article/pii/S0021999121007130
[2] https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2022MS003327
[3] https://www.sciencedirect.com/science/article/pii/S0021999119300233

[BFB] - not used by E3SM

### CODE REVIEW

**Comment by mark-petersen** [comment]
I tested with gcc 9.2 and intel 20.0 on Chrysalis, against the master branchpoint. The nightly suite passes and is bfb for all four: intel and gnu; debug and optimized. I tested this branch both with and without `USE_LTS=true` with the identical bfb result. I also tested with gnu on chicoma and cori, with the same results.

Note that `USE_LTS=true` allocates memory for four internal time levels, but LTS still needs to be turned on in the namelist with `config_time_integrator='LTS'`. Thus I tested with the compile flag `USE_LTS=true` but with the default `config_time_integrator` values, and I still get bfb match to master.

File: components/mpas-ocean/src/shared/mpas_ocn_vel_tidal_potential.F [hunk 1]
```diff
@@ -173,10 +175,10 @@ subroutine ocn_vel_tidal_potential_tend(ssh, surfacePressure, tend, err)!{{{
...
 
          potentialGrad = - gravity*invdcEdge* &
-            (  tidalPotEta(cell2) - tidalPotEta(cell1) &
-             + (1.0_RKIND - ssh_sal_on) * betaSelfAttrLoad *( (ssh(cell2) - ssh(cell1)) & 
-                                                      + rho0gInv*(surfacePressure(cell2) - surfacePressure(cell1))) &
-             + ssh_sal_on * (ssh_sal(cell2) - ssh_sal(cell1)) )
+            (  nonSshTermsSplit * tidalPotEta(cell2) - tidalPotEta(cell1) &
+             + sshTermSplit * (1.0_RKIND - ssh_sal_on) * betaSelfAttrLoad * ((ssh(cell2) - ssh(cell1)) & 
+             + nonSshTermsSplit * rho0gInv*(surfacePressure(cell2) - surfacePressure(cell1))) &
+             + nonSshTermsSplit * ssh_sal_on * (ssh_sal(cell2) - ssh_sal(cell1)) )
```
Comment (mark-petersen) [inline components/mpas-ocean/src/shared/mpas_ocn_vel_tidal_potential.F:181]: @gcapodag I discussed these lines with @philipwjones, and we decided that a separate pressure gradient option and separate loop is better than 1/0 coefficients for both clarity and performance.

Comment (gcapodag) [reply components/mpas-ocean/src/shared/mpas_ocn_vel_tidal_potential.F:181]: I agree, let's put an `if(config_time_integrator =='LTS')` before the`iEdge` loop you mentioned above, and also another one outside the `iEdge` loop in `src/shared/mpas_ocn_vel_tidal_potential.F`. If the time integration scheme is not LTS then we'll have the same loops, except with `sshTermSplit` and `nonSshTermsSplit` removed. Thanks for suggesting this change @mark-petersen @philipwjones !

**Comment by mark-petersen** [comment]
After the last two commits, I recompiled with both gnu and intel with LTS=false. Nightly suite (with no LTS) remains bfb with master using gnu optimized. So far so good...

**Comment by mark-petersen** [comment]
I then tested the hurricane Sandy storm surge as follows on cori:
```
MPAS_PATH=/global/homes/m/mpeterse/repos/E3SM/master/components/mpas-ocean
WORK_DIR=/global/cscratch1/sd/mpeterse/runs/230306_DEQU120at30cr10rr2_master

compass setup -t ocean/hurricane/DEQU120at30cr10rr2/mesh -w $WORK_DIR -p $MPAS_PATH
compass setup -t ocean/hurricane/DEQU120at30cr10rr2/init -w $WORK_DIR -p $MPAS_PATH
compass setup -t ocean/hurricane/DEQU120at30cr10rr2/sandy -w $WORK_DIR -p $MPAS_PATH
```

For some reason the pointwise stats are not working (see #5510). But I added global stats as follows:
```
namelist.ocean:
507 &AM_globalStats
508     config_AM_globalStats_enable = .true.

streams.ocean:
<stream name="globalStatsOutput"
        type="output"
        filename_template="globalStats.$Y-$M-$D_$h.$m.$s.nc"
        filename_interval="01-00-00_00:00:00"
        clobber_mode="truncate"
        packages="globalStatsAMPKG"
        output_interval="0000_01:00:00" >

    <var_array name="minGlobalStats"/>
    <var_array name="maxGlobalStats"/>
    <var_array name="sumGlobalStats"/>
    <var_array name="rmsGlobalStats"/>
    <var_array name="avgGlobalStats"/>
    <var name="xtime"/>
    <var name="daysSinceStartOfSim"/>
    <var name="volumeCellGlobal"/>
    <var name="CFLNumberGlobal"/>
</stream>
```

I want to test hurricane sandy on the following code configurations
1. master
2. lts branch, compiled with USE_LTS=false
3. lts branch, compiled with USE_LTS=true and all default flags (`config_time_integrator='RK4'`)
4. lts branch, compiled with USE_LTS=true and `config_time_integrator='LTS'`

I start with a short 5-day run using `config_stop_time = '2012-10-15_00:00:00'` and look at the avg kinetic energy at the end. This is on master:
```
cd /global/cscratch1/sd/mpeterse/runs/230306_DEQU120at30cr10rr2_master/ocean/hurricane/DEQU120at30cr10rr2/sandy/forward

$ ncdump -v xtime globalStats.2012-10-10_00.00.00.nc |tail -n 3
  "2012-10-14_23:00:00",
  "2012-10-15_00:00:00" ;
}
$ ncdump -v kineticEnergyCellAvg globalStats.2012-10-10_00.00.00.nc |tail -n 3
    0.000190995094378389, 0.000200727866937962, 0.000215410916170907,
    0.000228600070770075 ;
}
```

When I go to step 2 above, testing on this branch, the kinetic energy is completely different:
```
cori:forward$ pwd
/global/cscratch1/sd/mpeterse/runs/230307a_DEQU120at30cr10rr2_lts_noLTS_compile/ocean/hurricane/DEQU120at30cr10rr2/sandy/forward
cori:forward$ ncdump -v xtime globalStats.2012-10-10_00.00.00.nc |tail -n 3
  "2012-10-14_23:00:00",
  "2012-10-15_00:00:00" ;
}
cori:forward$ ncdump -v kineticEnergyCellAvg globalStats.2012-10-10_00.00.00.nc |tail -n 3
    6.86010426711228e-06, 6.96234837193462e-06, 7.10732816688074e-06,
    7.31631132029089e-06 ;
}
```
@gcapodag can you test this as well? This is puzzling because the nightly suite matches. So there must be a difference in the tidal forcing, or I made a mistake in setting up these tests. The low KE appears both before and after my last two commits.

**Comment by gcapodag** [comment]
@mark-petersen there was indeed an issue. I was calling `ocn_vel_tidal_potential_init` after the computation of the tidal forcing, but that routine is not like other init routines, because it also zeros out `tidalPotEta`. I took the chance to revise the code and get rid of all the extra init routines I was calling in `ocn_tend_vel` for LTS.

**Comment by mark-petersen** [comment]
@gcapodag, thank you! That fixed my previous problem. With your last commit f088d32, I recompiled on chrysalis and cori with both gnu and intel, debug and optimized, with USE_LTS=false and again with true. I ran the nightly regression suite with various combinations and they all match bfb with master.

I then proceeded with hurricane Sandy on cori with gnu optimized, with the following:

1. master
2. lts branch, compiled with USE_LTS=false
3. lts branch, compiled with USE_LTS=true and all default flags (config_time_integrator='RK4')
4. lts branch, compiled with USE_LTS=true and config_time_integrator='LTS'

I can now verify that steps 1, 2, and 3 match bfb as expected. Here are my steps on cori:
```
salloc -N 1  -t 30:00 --account=e3sm -C haswell
source /global/homes/m/mpeterse/repos/compass/master/load_dev_compass_*cori-haswell_gnu_*.sh
cd /global/cscratch1/sd/mpeterse/runs/230307b_DEQU120at30cr10rr2_lts_LTS_compile_ltsOFF/ocean/hurricane/DEQU120at30cr10rr2/sandy/forward
srun -n 64 ocean_model

ncdump -v kineticEnergyCellAvg globalStats.2012-10-10_00.00.00.nc |tail -n 3
    0.000190995094378389, 0.000200727866937962, 0.000215410916170907,
    0.000228600070770075 ;
```
Now proceeding to step 4, with `config_time_integrator='LTS'`!

**Comment by mark-petersen** [comment]
@gcapodag, one small issue is that every processor writes `Pressure type is: ssh_gradient` lines into the log file at every step:
```
  1245  Doing timestep 2012-10-10_00:05:25
  1246   Pressure type is: ssh_gradient
  1247   Pressure type is: ssh_gradient
  1248   Pressure type is: ssh_gradient
  1249   Pressure type is: ssh_gradient
  1250  Verifying that cells are not dry...
  1251  Minimum thickness is 9.99999996068664.
  1252  Done verifying that cells are wet.
```
We need to stop that, because writing to disk can be slow, and all cores will have file contention. It occurs here:
```
src/shared/mpas_ocn_vel_pressure_grad.F:677:         call mpas_log_write(' Pressure type is: ssh_gradient')
```
in the init routine. You are calling that every timestep, right? Is that intentional? It doesn't look like you would need to call this one every time: 
```
640    subroutine ocn_vel_pressure_grad_init(err)!{{{
```

**Comment by gcapodag** [comment]
@mark-petersen I forgot to remove that one too in my last commit. I pushed a commit now where it is removed. Thanks for catching that.

**Comment by mark-petersen** [comment]
The previous force push was to remove some submodules altered by mistake, with @gcapodag consent.

I can now run with LTS on. That is, step 4 on my list above:
4. lts branch, compiled with USE_LTS=true and config_time_integrator='LTS'

I tested this here on cori, with these flags. 
```
cd /global/cscratch1/sd/mpeterse/runs/230320d_DEQU120at30cr10rr2_lts_LTS_compile_ltsON_wInput/ocean/hurricane/DEQU120at30cr10rr2/sandy/forward

 25 &time_integration
 26     config_dt = '00:00:25'
 27     config_time_integrator = 'LTS'
 28 /
 29 &lts
 30     config_dt_scaling_LTS = 4
```
Note that I needed to link to @gcapodag input file in order to include the LTSRegions variable, and the graph.info files:
```
cori:forward$ pwd
/global/cscratch1/sd/mpeterse/runs/230320d_DEQU120at30cr10rr2_lts_LTS_compile_ltsON_wInput/ocean/hurricane/DEQU120at30cr10rr2/sandy/forward
cori:forward$ ls -lh input.nc graph*
lrwxrwxrwx 1 mpeterse mpeterse 116 Mar 20 14:09 graph.info -> /global/cscratch1/sd/gcapodag/e3sm_scratch/tidesTestCase/ocean/hurricane/DEQU120at30cr10rr2/sandy/forward/graph.info
lrwxrwxrwx 1 mpeterse mpeterse 126 Mar 20 14:09 graph.info.part.1024 -> /global/cscratch1/sd/gcapodag/e3sm_scratch/tidesTestCase/ocean/hurricane/DEQU120at30cr10rr2/sandy/forward/graph.info.part.1024
lrwxrwxrwx 1 mpeterse mpeterse 123 Mar 20 14:09 graph.info.part.2 -> /global/cscratch1/sd/gcapodag/e3sm_scratch/tidesTestCase/ocean/hurricane/DEQU120at30cr10rr2/sandy/forward/graph.info.part.2
lrwxrwxrwx 1 mpeterse mpeterse 124 Mar 20 14:09 graph.info.part.32 -> /global/cscratch1/sd/gcapodag/e3sm_scratch/tidesTestCase/ocean/hurricane/DEQU120at30cr10rr2/sandy/forward/graph.info.part.32
lrwxrwxrwx 1 mpeterse mpeterse 123 Mar 20 14:09 graph.info.part.4 -> /global/cscratch1/sd/gcapodag/e3sm_scratch/tidesTestCase/ocean/hurricane/DEQU120at30cr10rr2/sandy/forward/graph.info.part.4
lrwxrwxrwx 1 mpeterse mpeterse 120 Mar 20 14:09 graph.info.rk4 -> /global/cscratch1/sd/gcapodag/e3sm_scratch/tidesTestCase/ocean/hurricane/DEQU120at30cr10rr2/sandy/forward/graph.info.rk4
lrwxrwxrwx 1 mpeterse mpeterse 114 Mar 20 14:09 input.nc -> /global/cscratch1/sd/gcapodag/e3sm_scratch/tidesTestCase/ocean/hurricane/DEQU120at30cr10rr2/sandy/forward/input.nc

ncdump -v LTSRegion input.nc |tail
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
    1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2,
    2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 1, 1, 1, 1, 1, 2, 2, 1,
    2, 1, 1, 1, 2, 2, 2, 1, 2, 2, 1, 1, 2, 1, 1, 1, 1 ;
}
```
The simulation results are now similar but not exactly the same as with RK4:
```
(e3sm_unified_1.8.0_nompi) cori:forward$ ncdump -v xtime globalStats.2012-10-10_00.00.00.nc |tail -n 3
  "2012-10-14_23:00:00",
  "2012-10-15_00:00:00" ;
}
(e3sm_unified_1.8.0_nompi) cori:forward$ ncdump -v kineticEnergyCellAvg globalStats.2012-10-10_00.00.00.nc |tail -n 3
    0.000192273566155529, 0.000201966883262167, 0.000216625050727042,
    0.000229880707130802 ;
}
```

**Comment by gcapodag** [comment]
Thank you @mark-petersen for your help improving the initial code submission.

**Comment by jonbob** [comment]
@gcapodag -- this PR won't work in E3SM as-is. Do you want me to pass along necessary changes? Or can you make me a collaborator on your fork?

**Comment by gcapodag** [comment]
> @gcapodag -- this PR won't work in E3SM as-is. Do you want me to pass along necessary changes? Or can you make me a collaborator on your fork?

What is the issue? I just sent you the invite to become collaborator.

**Comment by jonbob** [comment]
E3SM uses cmake for the build, and you would have to add your new file to it. Also we use a set of semi-automated scripts to make the E3SM bld files consistent with Registry. You got most of it by hand, but there's at least one thing missing. I'll fix them and push the changes.

**Comment by gcapodag** [comment]
> @mark-petersen, @gcapodag - we don't really have conditional cppdefs setup in the mpas components, meaning nothing like a configure step that some components. I can try setting the USE_LTS cppdef in a testdef and see if that works? But this is not how we have typically handled mpas in E3SM

Hi @jonbob , I talked with @mark-petersen about this and he said he will get back to you on this matter. Thanks!

**Comment by gcapodag** [comment]
@mark-petersen the latest rebase is causing problems in the LTS compass test case. The forward mode is crashing. There are differences in the name list file for the forward mode:
1. Rayleigh damping before rebase:

> &Rayleigh_damping
>     **config_Rayleigh_friction = .false.**
>     config_Rayleigh_damping_coeff = 1.0e-4
>     config_Rayleigh_damping_depth_variable = .false.
>     config_Rayleigh_bottom_friction = .false.
>     config_Rayleigh_bottom_damping_coeff = 1.0e-4

After rebase:
> &Rayleigh_damping
>     config_Rayleigh_damping_coeff = 1.0e-4
>     config_Rayleigh_damping_depth_variable = .false.
>     config_Rayleigh_bottom_friction = .false.
>     config_Rayleigh_bottom_damping_coeff = 1.0e-4

2. Bottom drag before rebase:

> &bottom_drag
>     **config_use_implicit_bottom_drag = .false.**
>     **config_implicit_bottom_drag_coeff = 1.0e-3**
>     **config_use_implicit_bottom_roughness = .false.**
>     **config_use_implicit_bottom_drag_variable = .false.**
>     **config_use_implicit_bottom_drag_variable_mannings = .false.**
>     **config_use_explicit_bottom_drag = .true.**
>     config_explicit_bottom_drag_coeff = 1.0e-3
>     config_use_topographic_wave_drag = .true.
>     config_topographic_wave_drag_coeff = 5.0e-3

After rebase:

> &bottom_drag
>     **config_bottom_drag_mode = 'implicit'**
>     **config_implicit_bottom_drag_type = 'constant'**
>     **config_implicit_constant_bottom_drag_coeff = 1.0e-3**
>     **config_use_implicit_top_drag = .false.**
>     **config_implicit_top_drag_coeff = 1.0e-3**
>     config_explicit_bottom_drag_coeff = 1.0e-3
>     config_use_topographic_wave_drag = .true.
>     config_topographic_wave_drag_coeff = 5.0e-3

3. Vegetation drag before rebase:

> &vegetation_drag
>     config_use_vegetation_drag = .false.
>     **config_use_vegetation_manning_equation = .false.**
>     config_vegetation_drag_coefficient = 1.09

After rebase:

> &vegetation_drag
>     config_use_vegetation_drag = .false.
>     config_vegetation_drag_coefficient = 1.09

Also, a flag in the land ice fluxes has been renamed but we are not using them in LTS so it should not be a problem.

**Comment by gcapodag** [comment]
Added an error message (and associated interruption of the run) if `config_disable_tr_all_tend=false`, per @dengwirda 's request.

**Comment by xylar** [comment]
@gcapodag, your last commit contains changes to submodules, which should not be included here.  Could you hard-reset to the commit before your last one and try making just the file changes, making sure not to commit any submodule changes?

**Comment by gcapodag** [comment]
ouch sorry about that! will do

**Comment by gcapodag** [comment]
@jonbob @mark-petersen hello folks, is this PR ready to be finalized?

**Comment by jonbob** [comment]
@gcapodag -- we still don't have a solution for having the use of this dependent on a cppdef. It's not something we have otherwise done for any of the MPAS components in E3SM. Are you planning to primarily use this in standalone? Either way, I'll talk to @mark-petersen now that he's back from vacation and see what we can come up with

**Comment by mark-petersen** [comment]
I looked into altering the framework Registry parsing code here:
`components/mpas-framework/src/tools/registry/gen_inc.c`
in order to be able to specify the number of time levels with a namelist flag, or some way other than a compile flag.

Unfortunately, I've come to the conclusion that the number of time levels MUST be specified at compile time, and the Registry file is the natural place to do it. That is because time levels is not an array dimension at all (neither static or dynamic). The auto-generated Fortran code creates _multiple_ arrays for each variable, one for each time level. That is, during the text pre-processing step of the E3SM build, it runs this code:
`components/mpas-framework/src/tools/registry/gen_inc.c`
which processes this file:
`components/mpas-ocean/src/Registry.xml`
and auto-generates all variable declarations here:
`components/mpas-ocean/src/inc/structs_and_variables.inc`
which is included here for compiling:
`src/driver/mpas_ocn_core_interface.F:`
`724:      #include "../inc/structs_and_variables.inc"`

Here is an autogenerated declaration for a single variable in `structs_and_variables.inc`
<details>
<summary> code snippet </summary> <p>

```
! Define variable normalVelocity
      allocate(r2aPtr(2))

! Setting up time level 1
      r2aPtr(1) % fieldName = 'normalVelocity'
      r2aPtr(1) % outputFieldName = 'normalVelocity'
      r2aPtr(1) % isVarArray = .false.
      r2aPtr(1) % isDecomposed = .true.
      r2aPtr(1) % hasTimeDimension = .true.
      r2aPtr(1) % isPersistent = .true.
      r2aPtr(1) % isActive = .false.
! Setting up dimensions
      r2aPtr(1) % dimNames(1) = 'nVertLevels'
      r2aPtr(1) % dimNames(2) = 'nEdges'
      r2aPtr(1) % defaultValue = MPAS_REAL_FILLVAL ! defaultValue taking specified missing_value
      allocate(r2aPtr(1) % attLists(1))
      allocate(r2aPtr(1) % attLists(1) % attList)
      call mpas_add_att(r2aPtr(1) % attLists(1) % attList, 'units', 'm s^-1')
      call mpas_add_att(r2aPtr(1) % attLists(1) % attList, 'long_name', 'horizontal velocity, normal component to an edge')
      r2aPtr(1) % maskName = 'edgeMask'
      call mpas_add_att(r2aPtr(1) % attLists(1) % attList, 'missing_value_mask', 'edgeMask')
      call mpas_add_att(r2aPtr(1) % attLists(1) % attList, '_FillValue', MPAS_REAL_FILLVAL)
      r2aPtr(1) % missingValue = MPAS_REAL_FILLVAL
      r2aPtr(1) % block => block

! Setting up time level 2
      r2aPtr(2) % fieldName = 'normalVelocity'
      r2aPtr(2) % outputFieldName = 'normalVelocity'
      r2aPtr(2) % isVarArray = .false.
      r2aPtr(2) % isDecomposed = .true.
      r2aPtr(2) % hasTimeDimension = .true.
      r2aPtr(2) % isPersistent = .true.
      r2aPtr(2) % isActive = .false.
! Setting up dimensions
      r2aPtr(2) % dimNames(1) = 'nVertLevels'
      r2aPtr(2) % dimNames(2) = 'nEdges'
      r2aPtr(2) % defaultValue = MPAS_REAL_FILLVAL ! defaultValue taking specified missing_value
      allocate(r2aPtr(2) % attLists(1))
      allocate(r2aPtr(2) % attLists(1) % attList)
      call mpas_add_att(r2aPtr(2) % attLists(1) % attList, 'units', 'm s^-1')
      call mpas_add_att(r2aPtr(2) % attLists(1) % attList, 'long_name', 'horizontal velocity, normal component to an edge')
      r2aPtr(2) % maskName = 'edgeMask'
      call mpas_add_att(r2aPtr(2) % attLists(1) % attList, 'missing_value_mask', 'edgeMask')
      call mpas_add_att(r2aPtr(2) % attLists(1) % attList, '_FillValue', MPAS_REAL_FILLVAL)
      r2aPtr(2) % missingValue = MPAS_REAL_FILLVAL
      r2aPtr(2) % block => block

      r2aPtr(1) % isActive = .true.
      r2aPtr(2) % isActive = .true.
      call mpas_pool_add_field(newSubPool, 'normalVelocity', r2aPtr)
      call mpas_pool_add_field(block % allFields, 'normalVelocity', r2aPtr)
```
</details>

Since time levels are specified at compile time, each executable has a fixed number of time levels. That means a compile-time flag is the only way to choose the time levels, as the `USE_LTS` flag does here. I need to think further about the implications of this for running within E3SM.

**Comment by xylar** [comment]
@mark-petersen, you may be right that there's not a good way around hard-coding the number of time levels but the snippet you posted looks like it could fairly easily be turned into a for loop over the number of time leves.  Is it that this is one bridge too far for the level of changes you want to make to the code-generator? If so, I understand.

It will also be considerable work in E3SM to add support for multiple builds, so it's a trade off in terms of where to spend the effort.

**Comment by xylar** [comment]
> That is because time levels are not an array dimension at all (neither static or dynamic)

I don't think that's quite correct.  The time level is a dimension for the `r2aPtr`, and it could be allocated based on a dimension name instead of hard-coded to 2. It's just not analogous to any other array dimensions we currently use.

**Comment by mark-petersen** [comment]
@xylar thanks for your thoughts on this. You are right. The auto-generated code could instead be set up like this:
```
! Define variable normalVelocity
     allocate(r2aPtr(config_n_time_levels))

     for iTime = 1,config_n_time_levels
         r2aPtr(iTime) % fieldName = 'normalVelocity'
         r2aPtr(iTime) % outputFieldName = 'normalVelocity'
         ... etc
```
as long as I can get a namelist value `config_n_time_levels` into this subroutine, I think that would work (barring unforeseen problems...).

**Comment by rljacob** [comment]
notes: can go in now that #5766 is merged.

**Comment by gcapodag** [comment]
Hi all, I pushed @mark-petersen 's rebase to master. I tested with the previous version and it was BFB so green light on my end. Thanks.

**Comment by mark-petersen** [comment]
I compiled and tested the latest version in stand-alone with default flags (LTS off) with gnu on chicoma and intel on chrysalis, both debug and optimized. The nightly test suite matches bfb with master using optimized for both gnu and intel. It also passed the E3SM test: `SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.chrysalis_gnu`

**Comment by mark-petersen** [comment]
After discussions with @gcapodag and @jonbob, we have decided to classify LTS as an MPAS-Standalone feature for this PR, rather than an E3SM stealth feature. See the explanation in the second paragraph of this PR description. Because LTS is meant for short, single-layer coastal inundation simulations, it does not make sense to add an LTS test for multi-layer ocean domains, or for longer (multi-year) climate simulations.

**Comment by jonbob** [comment]
passes:
* SMS_D_Ld3.T62_oQU120.CMPASO-IAF.chrysalis_intel
* SMS_D_Ld1.ne30pg2_EC30to60E2r2.WCYCL1850.chrysalis_intel.allactive-wcprod
* ERS_Ld5.T62_oQU120.CMPASO-NYF.chrysalis_intel

merged to next

**Comment by jonbob** [comment]
merged to master

**Comment by gcapodag** [comment]
Fantastic! Thank you @mark-petersen and @jonbob for your work on this!

**Pull request merged**